src/raft/raft.rs

```diff
@@ -648,25 +696,39 @@ impl<T: Storage> Raft<T> {
     }
 
     fn campaign(&mut self, campaign_type: &[u8]) {
-        self.become_candidate();
+        let (vote_msg, term) = if campaign_type == CAMPAIGN_PRE_ELECTION {
+            self.become_precandidate();
+            // pre_vote RPCs are sent for next term before we've incremented self.term.
+            (MessageType::MsgRequestPreVote, self.term + 1)
+        } else {
+            self.become_candidate();
+            (MessageType::MsgRequestVote, self.term)
+        };
         let id = self.id;
-        let poll_res = self.poll(id, true);
-        if self.quorum() == poll_res {
-            self.become_leader();
+        if self.quorum() == self.poll(id, vote_resp_msg_type(vote_msg), true) {
+            // We won the election after voting for ourselves (which must mean that
+            // this is a single-node cluster). Advance to the next state.
+            if campaign_type == CAMPAIGN_PRE_ELECTION {
+                self.campaign(CAMPAIGN_ELECTION);
+            } else {
+                self.become_leader();
+            }
             return;
         }
         let ids: Vec<_> = self.prs.keys().cloned().collect();
         for id in ids {
             if id == self.id {
                 continue;
             }
-            info!("{} [logterm: {}, index: {}] sent vote request to {} at term {}",
+            info!("{} [logterm: {}, index: {}] sent {:?} request to {} at term {}",
```

Comment:
Implement trait `Display` for message type or add a string casting function for it so that the log message would be vivid.